cpp/dolfin/fem/dofs_permutation.cpp

```diff
@@ -429,233 +429,152 @@ hexahedron_rotations_and_reflection(const int volume_dofs, const int blocksize)
+  const int num_vertices_per_cell = mesh::num_cell_vertices(type);
+
+  // Find the total number of edges + faces + volumes
+  int entity_count = 0;
+  for (int dim = 1; dim < t_dim; ++dim)
+    entity_count += mesh::cell_num_entities(type, dim);
+  entity_count += 1;
+
+  // Get lists of vertices on each entity from the cell type
+  std::vector<std::pair<mesh::CellType, Eigen::Array<int, 1, Eigen::Dynamic>>>
+      entities(entity_count);
+  { // scope for j
+    int j = 0;
+    for (int dim = 1; dim < t_dim; ++dim)
+    {
+      auto vertices = mesh::get_entity_vertices(type, dim);
+      auto e_type = mesh::cell_entity_type(type, dim);
+      for (int i = 0; i < mesh::cell_num_entities(type, dim); ++i)
+      {
+        // Store the cell type and vertices associated with the ith
+        // entity of dimension dim
+        entities[j].first = e_type;
+        entities[j].second = vertices.row(i);
+        ++j;
+      }
+    }
+    // Add on the cell itself as an entity
+    entities[j].first = type;
+    entities[j].second.resize(num_vertices_per_cell);
+    for (int i = 0; i < num_vertices_per_cell; ++i)
+      entities[j].second[i] = i;
+    ++j;
+    assert(j == entity_count);
   }
 
-  return cell_orders;
-}
-//-----------------------------------------------------------------------------
-Eigen::Array<std::int8_t, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
-compute_ordering_interval(const mesh::Mesh& mesh)
-{
-  const int num_cells = mesh.num_entities(mesh.topology().dim());
-  const int num_permutations = get_num_permutations(mesh.cell_type());
+  // Set orders for each cell
   Eigen::Array<std::int8_t, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>
       cell_orders(num_cells, num_permutations);
-
-  // Set orders for each cell
   const std::vector<std::int64_t>& global_indices
       = mesh.topology().global_indices(0);
+  // Reserve memory to store global numbers of vertices of cell
+  std::vector<std::int32_t> cell_vs(num_vertices_per_cell);
   for (int cell_n = 0; cell_n < num_cells; ++cell_n)
   {
-    const mesh::MeshEntity cell(mesh, 1, cell_n);
-    const int* vertices = cell.entities(0);
-    const std::int64_t v0 = global_indices[vertices[0]];
-    const std::int64_t v1 = global_indices[vertices[1]];
+    // Get the cell and info about it
+    const mesh::MeshEntity cell(mesh, t_dim, cell_n);
+    const std::int32_t* local_vertices = cell.entities(0);
+    for (int i = 0; i < num_vertices_per_cell; ++i)
+      cell_vs[i] = global_indices[local_vertices[i]];
 
-    // Set the orders for the edge flip
-    cell_orders(cell_n, 0) = (v0 > v1);
-  }
+    int j = 0;
+    // iterate over the cell's entities
+    for (std::size_t e_n = 0; e_n < entities.size(); ++e_n)
+    {
+      // Look up the cell type of the entity, then calculate the orders
+      // for that entity type
+      auto cell_type = entities[e_n].first;
+      auto v = entities[e_n].second;
+      switch (cell_type)
```

Comment:
Not ideal to have a switch in the inner most loop. Is there another way?

Reply:
In 48ad08512e7171016f69ee6ad6931358b993c99b, I've moved the switch outside the loop and made it store the function needed rather than the cell type